gdx/src/com/badlogic/gdx/utils/ObjectMap.java

```diff
@@ -1,846 +1,779 @@
+	 *
+	 * @param key       a K key that will be checked for equality if a similar-seeming key is found
+	 * @param placement as calculated by {@link #place(Object)}, almost always with {@code place(key)}
+	 * @return the location in the key array of key, if found, or -1 if it was not found.
+	 */
+	int locateKey (final K key, final int placement) {
+		for (int i = placement; ; i = i + 1 & mask) {
+			// empty space is available
+			if (keyTable[i] == null) {
+				return -1;
+			}
+			if (key.equals(keyTable[i])) {
+				return i;
+			}
+		}
+	}
+
+	/**
+	 * Returns the old value associated with the specified key, or null.
+	 */
+	public V put (K key, V value) {
+		if (key == null)
+			throw new IllegalArgumentException("key cannot be null.");
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		int b = place(key);
+		int loc = locateKey(key, b);
+		// an identical key already exists
+		if (loc != -1) {
+			V tv = valueTable[loc];
+			valueTable[loc] = value;
+			return tv;
+		}
+		for (int i = b; ; i = (i + 1) & mask) {
+			// space is available so we insert and break (resize is later)
+			if (keyTable[i] == null) {
+				keyTable[i] = key;
+				valueTable[i] = value;
+				break;
+			}
+		}
+		if (++size >= threshold) {
+			resize(keyTable.length << 1);
+		}
+		return null;
+	}
+
+	//	public void putAll (ObjectMap<K, V> map) {
+//		ensureCapacity(map.size);
+//		for (Entry<K, V> entry : map)
+//			put(entry.key, entry.value);
+//	}
+	public void putAll (ObjectMap<K, V> map) {
+		ensureCapacity(map.size);
+		final K[] keyTable = map.keyTable;
+		final V[] valueTable = map.valueTable;
+		K k;
+		for (int i = 0, n = keyTable.length; i < n; i++) {
+			if ((k = keyTable[i]) != null)
+				put(k, valueTable[i]);
+		}
+	}
+
+	/**
+	 * Skips checks for existing keys.
+	 */
+	private void putResize (K key, V value) {
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		int b = place(key);
+		for (int i = b; ; i = (i + 1) & mask) {
+			// space is available so we insert and break (resize is later)
+			if (keyTable[i] == null) {
+				keyTable[i] = key;
+				valueTable[i] = value;
+				break;
+			}
+		}
+		if (++size >= threshold) {
+			resize(keyTable.length << 1);
+		}
+	}
+
+	/**
+	 * Returns the value for the specified key, or null if the key is not in the map.
+	 */
+	public V get (K key) {
+		final int loc = locateKey(key);
+		return loc == -1 ? null : valueTable[loc];
+	}
+
+	/**
+	 * Returns the value for the specified key, or the default value if the key is not in the map.
+	 */
+	public V get (K key, V defaultValue) {
+		final int loc = locateKey(key);
+		return loc == -1 ? defaultValue : valueTable[loc];
+	}
+
+	public V remove (K key) {
+		int loc = locateKey(key);
+		if (loc == -1) {
+			return null;
+		}
+		V oldValue = valueTable[loc];
+		while ((key = keyTable[loc + 1 & mask]) != null && (loc + 1 & mask) != place(key)) {
```

Comment:
avoid doubled calculations (loc + 1 & mask) => use a variable to calc the index

Reply:
Thanks for having me look at this code again, I may have spotted a bug.

Reply:
Maybe we should add some additional tests to make sure they are no other bugs?

Reply:
Welcome.

Reply:
That's part of the problem, this passed Apache Harmony's relevant tests and the bug never surfaced once, not even for a million removals. It would require very specific conditions to occur, mostly relating to hash collisions in the highest slots that wrap around to the earliest, and I'll double-check for other cases this could occur. Most other code used a different technique to look up across that border, because it wasn't removing; only remove() (the class method and also remove() in a nested Iterator class) removes.

Reply:
@tommyettinger can you come up with a test that covers "everything"? I think that's the main thing, it should work solidly.

Reply:
I don't believe an all-covering test for every usage of a map can exist, that's an infinitely-deep pitfall during TDD. I'll link the  tests I have run against; Harmony's tests are supposed to cover "everything" but obviously that's unreasonable. The adapted Harmony [tests for ObjectMap are here](https://github.com/tommyettinger/merry-ds/blob/master/src/test/java/ds/merry/ObjectMapTest.java#L444-L494) (highlighted range is for remove()). This is already a much higher bar than was set for the current ObjectMap, which doesn't seem to have any tests for most methods, and when it does, the tests are skeletal.

Reply:
@tommyettinger May basing on the hashing it would be nice to know how high is the chance for collisions is and maybe a few different read write delete combinations would be interesting + performance